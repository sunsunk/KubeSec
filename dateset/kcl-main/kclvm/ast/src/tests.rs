use crate::node_ref;
use crate::walker::MutSelfMutWalker;
use crate::{ast, ast::*};

/// Construct an AssignStmt node with assign_value as value
fn build_assign_node(attr_name: &str, assign_value: NodeRef<Expr>) -> NodeRef<Stmt> {
    let iden = node_ref!(Identifier {
        names: vec![Node::dummy_node(attr_name.to_string())],
        pkgpath: String::new(),
        ctx: ExprContext::Store
    });

    node_ref!(Stmt::Assign(AssignStmt {
        value: assign_value,
        targets: vec![iden],
        ty: None
    }))
}

fn get_dummy_assign_ast() -> ast::Node<ast::AssignStmt> {
    let filename = "main.k";
    let line = 1;
    let column = 1;
    let end_line = 1;
    let end_column = 2;
    ast::Node::new(
        ast::AssignStmt {
            targets: vec![Box::new(ast::Node::new(
                ast::Identifier {
                    names: vec![Node::dummy_node(String::from("a"))],
                    pkgpath: String::from(filename),
                    ctx: ast::ExprContext::Load,
                },
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            ))],
            value: Box::new(ast::Node::new(
                ast::Expr::StringLit(ast::StringLit {
                    is_long_string: false,
                    raw_value: String::from("s"),
                    value: String::from("s"),
                }),
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            )),
            ty: None,
        },
        String::from(filename),
        line,
        column,
        end_line,
        end_column,
    )
}

fn get_dummy_assign_binary_ast() -> ast::Node<ast::AssignStmt> {
    let filename = "main.k";
    let line = 1;
    let column = 1;
    let end_line = 1;
    let end_column = 2;
    ast::Node::new(
        ast::AssignStmt {
            targets: vec![Box::new(ast::Node::new(
                ast::Identifier {
                    names: vec![Node::dummy_node(String::from("a"))],
                    pkgpath: String::from(filename),
                    ctx: ast::ExprContext::Load,
                },
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            ))],
            value: Box::new(ast::Node::new(
                ast::Expr::Binary(ast::BinaryExpr {
                    op: ast::BinOp::Add,
                    left: Box::new(ast::Node::new(
                        ast::Expr::Identifier(ast::Identifier {
                            names: vec![Node::dummy_node(String::from("a"))],
                            pkgpath: String::from(filename),
                            ctx: ast::ExprContext::Load,
                        }),
                        String::from(filename),
                        line,
                        column,
                        end_line,
                        end_column,
                    )),
                    right: Box::new(ast::Node::new(
                        ast::Expr::Identifier(ast::Identifier {
                            names: vec![Node::dummy_node(String::from("a"))],
                            pkgpath: String::from(filename),
                            ctx: ast::ExprContext::Load,
                        }),
                        String::from(filename),
                        line,
                        column,
                        end_line,
                        end_column,
                    )),
                }),
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            )),
            ty: None,
        },
        String::from(filename),
        line,
        column,
        end_line,
        end_column,
    )
}

#[test]
fn test_ast_print_assign() {
    let assign_stmt = get_dummy_assign_ast();
    println!("{:?}", assign_stmt);
    let json_str = serde_json::to_string(&assign_stmt).unwrap();
    println!("{:?}", json_str);
}

#[test]
fn test_ast_print_assign_binary() {
    let assign_stmt = get_dummy_assign_binary_ast();
    println!("{:?}", assign_stmt);
    let json_str = serde_json::to_string(&assign_stmt).unwrap();
    println!("{:?}", json_str);
}

#[test]
fn test_mut_walker() {
    pub struct VarMutSelfMutWalker;
    impl<'ctx> MutSelfMutWalker<'ctx> for VarMutSelfMutWalker {
        fn walk_identifier(&mut self, identifier: &'ctx mut ast::Identifier) {
            if identifier.names[0].node == "a" {
                let id_mut = identifier.names.get_mut(0).unwrap();
                id_mut.node = "x".to_string();
            }
        }
    }
    let mut assign_stmt = get_dummy_assign_ast();
    VarMutSelfMutWalker {}.walk_assign_stmt(&mut assign_stmt.node);
    assert_eq!(assign_stmt.node.targets[0].node.names[0].node, "x")
}

#[test]
fn test_try_from_for_stringlit() {
    let str_lit = ast::StringLit::try_from("test_str".to_string()).unwrap();
    let json_str = serde_json::to_string(&str_lit).unwrap();

    let str_expected =
        r#"{"is_long_string":false,"raw_value":"\"test_str\"","value":"test_str"}"#.to_string();
    assert_eq!(str_expected, json_str);
}

#[test]
fn test_try_from_for_nameconstant() {
    let name_cons = ast::NameConstant::try_from(true).unwrap();
    let json_str = serde_json::to_string(&name_cons).unwrap();
    assert_eq!("\"True\"", json_str);

    let name_cons = ast::NameConstant::try_from(false).unwrap();
    let json_str = serde_json::to_string(&name_cons).unwrap();
    assert_eq!("\"False\"", json_str);
}

#[test]
fn test_filter_schema_with_no_schema() {
    let ast_mod = Module {
        filename: "".to_string(),
        pkg: "".to_string(),
        doc: Some(node_ref!("".to_string())),
        name: "".to_string(),
        body: vec![],
        comments: vec![],
    };
    let schema_stmts = ast_mod.filter_schema_stmt_from_module();
    assert_eq!(schema_stmts.len(), 0);
}

#[test]
fn test_filter_schema_with_one_schema() {
    let mut ast_mod = Module {
        filename: "".to_string(),
        pkg: "".to_string(),
        doc: Some(node_ref!("".to_string())),
        name: "".to_string(),
        body: vec![],
        comments: vec![],
    };
    let mut gen_schema_stmts = gen_schema_stmt(1);
    ast_mod.body.append(&mut gen_schema_stmts);
    let schema_stmts = ast_mod.filter_schema_stmt_from_module();
    assert_eq!(schema_stmts.len(), 1);
    assert_eq!(schema_stmts[0].node.name.node, "schema_stmt_0".to_string());
}

#[test]
fn test_filter_schema_with_mult_schema() {
    let mut ast_mod = Module {
        filename: "".to_string(),
        pkg: "".to_string(),
        doc: Some(node_ref!("".to_string())),
        name: "".to_string(),
        body: vec![],
        comments: vec![],
    };
    let mut gen_schema_stmts = gen_schema_stmt(10);
    ast_mod.body.append(&mut gen_schema_stmts);
    let schema_stmts = ast_mod.filter_schema_stmt_from_module();
    assert_eq!(schema_stmts.len(), 10);
    for i in 0..10 {
        assert_eq!(
            schema_stmts[i].node.name.node,
            "schema_stmt_".to_string() + &i.to_string()
        )
    }
}

#[test]
fn test_build_assign_stmt() {
    let test_expr = node_ref!(ast::Expr::Identifier(Identifier {
        names: vec![
            Node::dummy_node("name1".to_string()),
            Node::dummy_node("name2".to_string())
        ],
        pkgpath: "test".to_string(),
        ctx: ast::ExprContext::Load
    }));
    let assgin_stmt = build_assign_node("test_attr_name", test_expr);

    if let ast::Stmt::Assign(ref assign) = assgin_stmt.node {
        if let ast::Expr::Identifier(ref iden) = &assign.value.node {
            assert_eq!(iden.names.len(), 2);
            assert_eq!(iden.names[0].node, "name1".to_string());
            assert_eq!(iden.names[1].node, "name2".to_string());
            assert_eq!(iden.pkgpath, "test".to_string());
            match iden.ctx {
                ast::ExprContext::Load => {}
                _ => {
                    assert!(false);
                }
            }
        } else {
            assert!(false);
        }
    } else {
        assert!(false);
    }
}

fn gen_schema_stmt(count: i32) -> Vec<NodeRef<ast::Stmt>> {
    let mut schema_stmts = Vec::new();
    for c in 0..count {
        schema_stmts.push(node_ref!(ast::Stmt::Schema(SchemaStmt {
            doc: Some(node_ref!("".to_string())),
            name: node_ref!("schema_stmt_".to_string() + &c.to_string()),
            parent_name: None,
            for_host_name: None,
            is_mixin: false,
            is_protocol: false,
            args: None,
            mixins: vec![],
            body: vec![],
            decorators: vec![],
            checks: vec![],
            index_signature: None
        })))
    }
    schema_stmts
}
