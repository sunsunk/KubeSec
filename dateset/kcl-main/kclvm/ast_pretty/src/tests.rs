use std::path::{Path, PathBuf};

use super::print_ast_module;
use kclvm_parser::parse_file_force_errors;
use pretty_assertions::assert_eq;

const FILE_INPUT_SUFFIX: &str = ".input";
const FILE_OUTPUT_SUFFIX: &str = ".output";
const TEST_CASES: &[&str] = &[
    "arguments",
    "empty",
    "if_stmt",
    "import",
    "unary",
    "codelayout",
    "collection_if",
    "comment",
    "index_sign",
    "joined_str",
    "lambda",
    "orelse",
    "quant",
    "rule",
    "str",
    "type_alias",
    "unification",
];

fn read_data(data_name: &str) -> (String, String) {
    let mut filename = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    filename.push(
        Path::new("src")
            .join("test_data")
            .join(format!("{}{}", data_name, FILE_INPUT_SUFFIX))
            .display()
            .to_string(),
    );

    let module = parse_file_force_errors(filename.to_str().unwrap(), None);

    let mut filename_expect = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    filename_expect.push(
        Path::new("src")
            .join("test_data")
            .join(format!("{}{}", data_name, FILE_OUTPUT_SUFFIX))
            .display()
            .to_string(),
    );
    (
        print_ast_module(&module.unwrap()),
        std::fs::read_to_string(filename_expect.to_str().unwrap()).unwrap(),
    )
}

#[test]
fn test_ast_printer() {
    for case in TEST_CASES {
        let (data_input, data_output) = read_data(case);

        #[cfg(target_os = "windows")]
        let data_output = data_output.replace("\r\n", "\n");

        assert_eq!(data_input, data_output, "Test failed on {}", case);
    }
}
