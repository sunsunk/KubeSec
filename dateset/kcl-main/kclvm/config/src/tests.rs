use kclvm_version as version;
use std::{
    collections::HashMap,
    env,
    fs::{self, File},
    path::{Path, PathBuf},
};

use crate::{
    cache::{load_pkg_cache, save_pkg_cache, CacheOption},
    modfile::{get_vendor_home, KCL_PKG_PATH},
};

#[test]
fn test_vendor_home() {
    env::set_var(KCL_PKG_PATH, "test_vendor_home");
    assert_eq!(get_vendor_home(), "test_vendor_home");
    env::remove_var(KCL_PKG_PATH);

    #[cfg(target_os = "windows")]
    let root_dir = env::var("USERPROFILE").unwrap();
    #[cfg(not(target_os = "windows"))]
    let root_dir = env::var("HOME").unwrap();

    let kpm_home = PathBuf::from(root_dir)
        .join(".kcl")
        .join("kpm")
        .canonicalize()
        .unwrap();
    assert_eq!(get_vendor_home(), kpm_home.display().to_string())
}

#[test]
fn test_pkg_cache() {
    let root = PathBuf::from("./src/testdata/test_cache/")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    let mut external_pkgs = HashMap::new();
    external_pkgs.insert(
        "test_vendor".to_string(),
        "./src/testdata/test_vendor".to_string(),
    );

    let lock_path = Path::new(&root)
        .join(".kclvm/cache")
        .join(format!("{}-{}", version::VERSION, version::CHECK_SUM))
        .join("test_target");

    fs::create_dir_all(lock_path.clone()).unwrap();
    File::create(lock_path.join("test_vendor.lock")).unwrap();

    save_pkg_cache(
        &root,
        "test_target",
        "test_vendor",
        "test_data",
        CacheOption::default(),
        &external_pkgs,
    )
    .unwrap();

    assert_eq!(
        load_pkg_cache(
            &root,
            "test_target",
            "test_vendor",
            CacheOption::default(),
            &external_pkgs,
        ),
        Some("test_data".to_string())
    )
}
