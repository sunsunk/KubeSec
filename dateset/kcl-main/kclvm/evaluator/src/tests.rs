use crate::Evaluator;
use kclvm_loader::{load_packages, LoadPackageOptions};
use kclvm_parser::LoadProgramOptions;

#[macro_export]
macro_rules! evaluator_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            let p = load_packages(&LoadPackageOptions {
                paths: vec!["test.k".to_string()],
                load_opts: Some(LoadProgramOptions {
                    k_code_list: vec![$src.to_string()],
                    ..Default::default()
                }),
                load_builtin: false,
                ..Default::default()
            })
            .unwrap();
            let evaluator = Evaluator::new(&p.program);
            insta::assert_snapshot!(format!("{}", evaluator.run().unwrap().1));
        }
    };
}

evaluator_snapshot! {expr_stmt_0, "1"}
evaluator_snapshot! {expr_stmt_1, "2.0"}
evaluator_snapshot! {expr_stmt_2, "True"}
evaluator_snapshot! {expr_stmt_3, r#"None"#}
evaluator_snapshot! {expr_stmt_4, r#"[1, 2, 3]"#}
evaluator_snapshot! {expr_stmt_5, r#"{k = "v"}"#}

evaluator_snapshot! {assign_stmt_0, "a = 1"}
evaluator_snapshot! {assign_stmt_1, "a = 1 + 1"}
evaluator_snapshot! {assign_stmt_2, "a = (1 + 2)"}
evaluator_snapshot! {assign_stmt_3, r#"a = 1
b = a + 1
"#}
evaluator_snapshot! {assign_stmt_4, r#"a: int = 1
b: int = a + 1
"#}

evaluator_snapshot! {aug_assign_stmt_0, r#"_a = 1
_a += 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_1, r#"_a = 1
_a -= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_2, r#"_a = 1
_a *= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_3, r#"_a = 2
_a /= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_4, r#"_a = 3
_a %= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_5, r#"_a = 3
_a **= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_6, r#"_a = 3
_a <<= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_7, r#"_a = 3
_a >>= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_8, r#"_a = 3
_a |= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_9, r#"_a = 3
_a ^= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_10, r#"_a = 3
_a &= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_11, r#"_a = 3
_a //= 2
a = _a
"#}

evaluator_snapshot! {assert_stmt_0, r#"assert True, "msg"
a = 1
"#}

evaluator_snapshot! {assert_stmt_1, r#"assert False if False, "msg"
a = 1
"#}

evaluator_snapshot! {if_stmt_0, r#"if True:
    a = 1
else:
    b = 2
"#}
evaluator_snapshot! {if_stmt_1, r#"if False:
    a = 1
else:
    b = 2
"#}
evaluator_snapshot! {if_stmt_3, r#"if False:
    a = 1
elif True:
    b = 2
else:
    c = 3
"#}
evaluator_snapshot! {if_stmt_4, r#"if False:
    a = 1
elif False:
    b = 2
else:
    c = 3
"#}
evaluator_snapshot! {if_stmt_5, r#"if False:
    a = 1
else:
    if True:
        b = 2
    else:
        c = 3
"#}

evaluator_snapshot! {import_stmt_0, r#"import math
a = 1
"#}
evaluator_snapshot! {import_stmt_1, r#"import math
import math
b = 2
"#}

evaluator_snapshot! {quant_expr_0, r#"b = all a in [1, 2, 3] {
    a > 0
}
"#}
evaluator_snapshot! {quant_expr_1, r#"b = any a in [1, 2, 3] {
    a > 2
}
"#}
evaluator_snapshot! {quant_expr_2, r#"b = all a in [1, 2, 3] {
    a > 5
}
"#}
evaluator_snapshot! {quant_expr_3, r#"b = any a in [1, 2, 3] {
    a > 5
}
"#}
evaluator_snapshot! {quant_expr_4, r#"b = map a in [1, 2, 3] {
    a + 1
}
"#}
evaluator_snapshot! {quant_expr_5, r#"b = filter a in [1, 2, 3] {
    a > 1
}
"#}
evaluator_snapshot! {quant_expr_6, r#"b = map a in [1, 2, 3] {
    a ** 2
}
"#}
evaluator_snapshot! {quant_expr_7, r#"b = filter a in [1, 2, 3] {
    a == 0
}
"#}

evaluator_snapshot! {if_expr_0, r#"a = 1 if True else 0"#}
evaluator_snapshot! {if_expr_1, r#"a = 1 if False else 0"#}
evaluator_snapshot! {if_expr_2, r#"a = 1 if False else 0 if False else 2"#}

evaluator_snapshot! {unary_expr_0, r#"a = +1"#}
evaluator_snapshot! {unary_expr_1, r#"a = -1"#}
evaluator_snapshot! {unary_expr_2, r#"a = ~1"#}
evaluator_snapshot! {unary_expr_3, r#"a = not None"#}

evaluator_snapshot! {binary_expr_0, r#"a = 1 + 1 * 2 - 4"#}
evaluator_snapshot! {binary_expr_1, r#"a = None or {}
b = [] and {}
"#}

evaluator_snapshot! {selector_expr_0, r#"a = {k = "v"}.k
b = {k = "v"}?.k
c = None?.k
"#}
evaluator_snapshot! {selector_expr_1, r#"a = [1, 2, 3]?[::-1]
b = a?[-1]
c = a?[0]
d = None?[1]
"#}

evaluator_snapshot! {subscript_expr_0, r#"a = [1, 2, 3][::-1]
b = a[-1]
c = a[0]
"#}
evaluator_snapshot! {subscript_expr_1, r#"a = [1, 2, 3]?[::-1]
b = a?[-1]
c = a?[0]
d = None?[1]
"#}

evaluator_snapshot! {compare_expr_0, r#"a = 1 < 10
b = 1 < 10 < 100
c = 1 > 10 > 100
d = 1 < 10 > 100
"#}

evaluator_snapshot! {paren_expr_0, r#"a = 2 * (1 + 1)
b = (((1 + 1))) * 2
"#}

evaluator_snapshot! {list_expr_0, r#"a = [1, 2, 3]
b = [1, if True: 2, 3]
c = [1, if False: 2, 3]
d = [1, *[2, 3]]
"#}

evaluator_snapshot! {dict_expr_0, r#"a = {k1 = "v1", k2 = "v2"}
b = {k1 = "v1", if True: k2 = "v2"}
c = {k1 = "v1", if False: k2 = "v2"}
d = {k1 = "v1", **{k2 = "v2"}}
"#}

evaluator_snapshot! {loop_0, r#"a = [i ** 2 for i in [1, 2, 3]]"#}
evaluator_snapshot! {loop_1, r#"a = [i + j for i in [1, 2, 3] for j in [1, 2, 3] if i < j]"#}

evaluator_snapshot! {literal_0, r#"longStringStartWithNewline = """\
This is the second line
This is the third line
"""
"#}
evaluator_snapshot! {literal_1, r#"a = {k = "v"}
b = "${a: #json}"
"#}
evaluator_snapshot! {literal_2, r#"a = 1Mi
b = 2K
"#}

evaluator_snapshot! {lambda_0, r#"f = lambda x {x * 2}
a = f(1)
b = f(2)
"#}
evaluator_snapshot! {lambda_1, r#"a = lambda x {x * 2}(1)
b = lambda x {x * 2}(2)
"#}
evaluator_snapshot! {lambda_2, r#"import math
a = math.log(10)
b = len("abc")
c = len([1, 2])
"#}
evaluator_snapshot! {lambda_3, r#"
x = lambda {
    a = 1
    lambda {
        a + 1
    }()
}()
"#}
evaluator_snapshot! {lambda_4, r#"
x = lambda {
    a = 1
    b = 2
    lambda x {
        a + b + x
    }(3)
}()
"#}
evaluator_snapshot! {lambda_5, r#"
func = lambda config: {str:} {
    x = 1
    lambda {
        y = 1
        lambda {
            z = 1
            lambda {
                {value = x + y + z + config.key}
            }()
        }()
    }()
}

x = func({key = 1})
"#}

evaluator_snapshot! {schema_0, r#"
schema Person:
    name: str = "Alice"
    age: int = 10

alice = Person {}
bob = Person {
    name = "Bob"
    age = 18
}
"#}
