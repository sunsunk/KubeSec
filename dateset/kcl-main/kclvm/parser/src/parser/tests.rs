use crate::parse_file_force_errors;
use regex::Regex;

#[test]
fn test_parse_file_not_found() {
    match parse_file_force_errors("The file path is invalid", None) {
        Ok(_) => {
            panic!("unreachable")
        }
        Err(err_msg) => {
            assert!(
                Regex::new(r"^Failed to load KCL file 'The file path is invalid'. Because.*")
                    .unwrap()
                    .is_match(&err_msg.to_string())
            );
        }
    }
}
