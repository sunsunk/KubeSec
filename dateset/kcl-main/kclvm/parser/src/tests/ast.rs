use crate::tests::parse_file_ast_json_snapshot;

parse_file_ast_json_snapshot!(
    schema_stmt,
    "hello.k",
    r####"
schema TestBool:
    []
    [str    ]: int
    [a: str]: int
    [a: ...str]: int
    [...str]: int
    a: int
    b?: str
    c: int = 0
    d?: str = ""

    [a]
    [a, b, c]
    [
        1
    ]
    [
        a
    ]
    [a for a in [1, 2, 3]]
    [
        a for a in [1, 2, 3]
    ]

    check:
        a > 1, "msg"
        name not None, "we fail here"
        "####
);
parse_file_ast_json_snapshot!(assign_stmt, "hello.k", r####"a=123"####);
parse_file_ast_json_snapshot!(
    if_stmt_0,
    "hello.k",
    r####"
a = 10
b = 12
_condition = 0
if a == 11 or b == 13: _condition = 1
elif a == 10 and b == 12: _condition = 2
condition = _condition
        "####
);
parse_file_ast_json_snapshot!(
    if_stmt_1,
    "hello.k",
    r####"
data2 = {
    **{key = "value1"}
    if a == 123: if b == 456: key = "value2"
}
    "####
);
parse_file_ast_json_snapshot!(
    basic_stmt,
    "hello.k",
    r####"
# comment1
a = 1
# comment22
b = 2
# comment333
c = 3 # comment4444
    "####
);
