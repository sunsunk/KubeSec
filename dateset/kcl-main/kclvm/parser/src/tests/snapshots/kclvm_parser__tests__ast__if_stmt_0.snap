---
source: parser/src/tests/ast.rs
expression: "crate::tests::parsing_file_ast_json(\"hello.k\",\n    r####\"\na = 10\nb = 12\n_condition = 0\nif a == 11 or b == 13: _condition = 1\nelif a == 10 and b == 12: _condition = 2\ncondition = _condition\n        \"####)"
---
{
  "filename": "hello.k",
  "pkg": "__main__",
  "doc": null,
  "name": "__main__",
  "body": [
    {
      "node": {
        "type": "Assign",
        "targets": [
          {
            "node": {
              "names": [
                {
                  "node": "a",
                  "filename": "hello.k",
                  "line": 2,
                  "column": 0,
                  "end_line": 2,
                  "end_column": 1
                }
              ],
              "pkgpath": "",
              "ctx": "Store"
            },
            "filename": "hello.k",
            "line": 2,
            "column": 0,
            "end_line": 2,
            "end_column": 1
          }
        ],
        "value": {
          "node": {
            "type": "NumberLit",
            "binary_suffix": null,
            "value": {
              "type": "Int",
              "value": 10
            }
          },
          "filename": "hello.k",
          "line": 2,
          "column": 4,
          "end_line": 2,
          "end_column": 6
        },
        "ty": null
      },
      "filename": "hello.k",
      "line": 2,
      "column": 0,
      "end_line": 2,
      "end_column": 6
    },
    {
      "node": {
        "type": "Assign",
        "targets": [
          {
            "node": {
              "names": [
                {
                  "node": "b",
                  "filename": "hello.k",
                  "line": 3,
                  "column": 0,
                  "end_line": 3,
                  "end_column": 1
                }
              ],
              "pkgpath": "",
              "ctx": "Store"
            },
            "filename": "hello.k",
            "line": 3,
            "column": 0,
            "end_line": 3,
            "end_column": 1
          }
        ],
        "value": {
          "node": {
            "type": "NumberLit",
            "binary_suffix": null,
            "value": {
              "type": "Int",
              "value": 12
            }
          },
          "filename": "hello.k",
          "line": 3,
          "column": 4,
          "end_line": 3,
          "end_column": 6
        },
        "ty": null
      },
      "filename": "hello.k",
      "line": 3,
      "column": 0,
      "end_line": 3,
      "end_column": 6
    },
    {
      "node": {
        "type": "Assign",
        "targets": [
          {
            "node": {
              "names": [
                {
                  "node": "_condition",
                  "filename": "hello.k",
                  "line": 4,
                  "column": 0,
                  "end_line": 4,
                  "end_column": 10
                }
              ],
              "pkgpath": "",
              "ctx": "Store"
            },
            "filename": "hello.k",
            "line": 4,
            "column": 0,
            "end_line": 4,
            "end_column": 10
          }
        ],
        "value": {
          "node": {
            "type": "NumberLit",
            "binary_suffix": null,
            "value": {
              "type": "Int",
              "value": 0
            }
          },
          "filename": "hello.k",
          "line": 4,
          "column": 13,
          "end_line": 4,
          "end_column": 14
        },
        "ty": null
      },
      "filename": "hello.k",
      "line": 4,
      "column": 0,
      "end_line": 4,
      "end_column": 14
    },
    {
      "node": {
        "type": "If",
        "body": [
          {
            "node": {
              "type": "Assign",
              "targets": [
                {
                  "node": {
                    "names": [
                      {
                        "node": "_condition",
                        "filename": "hello.k",
                        "line": 5,
                        "column": 23,
                        "end_line": 5,
                        "end_column": 33
                      }
                    ],
                    "pkgpath": "",
                    "ctx": "Store"
                  },
                  "filename": "hello.k",
                  "line": 5,
                  "column": 23,
                  "end_line": 5,
                  "end_column": 33
                }
              ],
              "value": {
                "node": {
                  "type": "NumberLit",
                  "binary_suffix": null,
                  "value": {
                    "type": "Int",
                    "value": 1
                  }
                },
                "filename": "hello.k",
                "line": 5,
                "column": 36,
                "end_line": 5,
                "end_column": 37
              },
              "ty": null
            },
            "filename": "hello.k",
            "line": 5,
            "column": 23,
            "end_line": 5,
            "end_column": 37
          }
        ],
        "cond": {
          "node": {
            "type": "Binary",
            "left": {
              "node": {
                "type": "Compare",
                "left": {
                  "node": {
                    "type": "Identifier",
                    "names": [
                      {
                        "node": "a",
                        "filename": "hello.k",
                        "line": 5,
                        "column": 3,
                        "end_line": 5,
                        "end_column": 4
                      }
                    ],
                    "pkgpath": "",
                    "ctx": "Load"
                  },
                  "filename": "hello.k",
                  "line": 5,
                  "column": 3,
                  "end_line": 5,
                  "end_column": 4
                },
                "ops": [
                  "Eq"
                ],
                "comparators": [
                  {
                    "node": {
                      "type": "NumberLit",
                      "binary_suffix": null,
                      "value": {
                        "type": "Int",
                        "value": 11
                      }
                    },
                    "filename": "hello.k",
                    "line": 5,
                    "column": 8,
                    "end_line": 5,
                    "end_column": 10
                  }
                ]
              },
              "filename": "hello.k",
              "line": 5,
              "column": 3,
              "end_line": 5,
              "end_column": 21
            },
            "op": "Or",
            "right": {
              "node": {
                "type": "Compare",
                "left": {
                  "node": {
                    "type": "Identifier",
                    "names": [
                      {
                        "node": "b",
                        "filename": "hello.k",
                        "line": 5,
                        "column": 14,
                        "end_line": 5,
                        "end_column": 15
                      }
                    ],
                    "pkgpath": "",
                    "ctx": "Load"
                  },
                  "filename": "hello.k",
                  "line": 5,
                  "column": 14,
                  "end_line": 5,
                  "end_column": 15
                },
                "ops": [
                  "Eq"
                ],
                "comparators": [
                  {
                    "node": {
                      "type": "NumberLit",
                      "binary_suffix": null,
                      "value": {
                        "type": "Int",
                        "value": 13
                      }
                    },
                    "filename": "hello.k",
                    "line": 5,
                    "column": 19,
                    "end_line": 5,
                    "end_column": 21
                  }
                ]
              },
              "filename": "hello.k",
              "line": 5,
              "column": 14,
              "end_line": 5,
              "end_column": 21
            }
          },
          "filename": "hello.k",
          "line": 5,
          "column": 3,
          "end_line": 5,
          "end_column": 21
        },
        "orelse": [
          {
            "node": {
              "type": "If",
              "body": [
                {
                  "node": {
                    "type": "Assign",
                    "targets": [
                      {
                        "node": {
                          "names": [
                            {
                              "node": "_condition",
                              "filename": "hello.k",
                              "line": 6,
                              "column": 26,
                              "end_line": 6,
                              "end_column": 36
                            }
                          ],
                          "pkgpath": "",
                          "ctx": "Store"
                        },
                        "filename": "hello.k",
                        "line": 6,
                        "column": 26,
                        "end_line": 6,
                        "end_column": 36
                      }
                    ],
                    "value": {
                      "node": {
                        "type": "NumberLit",
                        "binary_suffix": null,
                        "value": {
                          "type": "Int",
                          "value": 2
                        }
                      },
                      "filename": "hello.k",
                      "line": 6,
                      "column": 39,
                      "end_line": 6,
                      "end_column": 40
                    },
                    "ty": null
                  },
                  "filename": "hello.k",
                  "line": 6,
                  "column": 26,
                  "end_line": 6,
                  "end_column": 40
                }
              ],
              "cond": {
                "node": {
                  "type": "Binary",
                  "left": {
                    "node": {
                      "type": "Compare",
                      "left": {
                        "node": {
                          "type": "Identifier",
                          "names": [
                            {
                              "node": "a",
                              "filename": "hello.k",
                              "line": 6,
                              "column": 5,
                              "end_line": 6,
                              "end_column": 6
                            }
                          ],
                          "pkgpath": "",
                          "ctx": "Load"
                        },
                        "filename": "hello.k",
                        "line": 6,
                        "column": 5,
                        "end_line": 6,
                        "end_column": 6
                      },
                      "ops": [
                        "Eq"
                      ],
                      "comparators": [
                        {
                          "node": {
                            "type": "NumberLit",
                            "binary_suffix": null,
                            "value": {
                              "type": "Int",
                              "value": 10
                            }
                          },
                          "filename": "hello.k",
                          "line": 6,
                          "column": 10,
                          "end_line": 6,
                          "end_column": 12
                        }
                      ]
                    },
                    "filename": "hello.k",
                    "line": 6,
                    "column": 5,
                    "end_line": 6,
                    "end_column": 24
                  },
                  "op": "And",
                  "right": {
                    "node": {
                      "type": "Compare",
                      "left": {
                        "node": {
                          "type": "Identifier",
                          "names": [
                            {
                              "node": "b",
                              "filename": "hello.k",
                              "line": 6,
                              "column": 17,
                              "end_line": 6,
                              "end_column": 18
                            }
                          ],
                          "pkgpath": "",
                          "ctx": "Load"
                        },
                        "filename": "hello.k",
                        "line": 6,
                        "column": 17,
                        "end_line": 6,
                        "end_column": 18
                      },
                      "ops": [
                        "Eq"
                      ],
                      "comparators": [
                        {
                          "node": {
                            "type": "NumberLit",
                            "binary_suffix": null,
                            "value": {
                              "type": "Int",
                              "value": 12
                            }
                          },
                          "filename": "hello.k",
                          "line": 6,
                          "column": 22,
                          "end_line": 6,
                          "end_column": 24
                        }
                      ]
                    },
                    "filename": "hello.k",
                    "line": 6,
                    "column": 17,
                    "end_line": 6,
                    "end_column": 24
                  }
                },
                "filename": "hello.k",
                "line": 6,
                "column": 5,
                "end_line": 6,
                "end_column": 24
              },
              "orelse": []
            },
            "filename": "hello.k",
            "line": 6,
            "column": 0,
            "end_line": 7,
            "end_column": 0
          }
        ]
      },
      "filename": "hello.k",
      "line": 5,
      "column": 0,
      "end_line": 7,
      "end_column": 0
    },
    {
      "node": {
        "type": "Assign",
        "targets": [
          {
            "node": {
              "names": [
                {
                  "node": "condition",
                  "filename": "hello.k",
                  "line": 7,
                  "column": 0,
                  "end_line": 7,
                  "end_column": 9
                }
              ],
              "pkgpath": "",
              "ctx": "Store"
            },
            "filename": "hello.k",
            "line": 7,
            "column": 0,
            "end_line": 7,
            "end_column": 9
          }
        ],
        "value": {
          "node": {
            "type": "Identifier",
            "names": [
              {
                "node": "_condition",
                "filename": "hello.k",
                "line": 7,
                "column": 12,
                "end_line": 7,
                "end_column": 22
              }
            ],
            "pkgpath": "",
            "ctx": "Load"
          },
          "filename": "hello.k",
          "line": 7,
          "column": 12,
          "end_line": 7,
          "end_column": 22
        },
        "ty": null
      },
      "filename": "hello.k",
      "line": 7,
      "column": 0,
      "end_line": 7,
      "end_column": 22
    }
  ],
  "comments": []
}
